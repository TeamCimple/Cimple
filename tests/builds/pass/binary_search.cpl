struct BinaryTree {
	struct BinaryTree *left;
	struct BinaryTree *right;
	int val;

	BinaryTree(int val_to_set) {
		val = val_to_set;
		left = NIL;
		right = NIL;
	}

	~BinaryTree() {
		printf("Cleaning binary tree\n");
		if (left != NIL) {
			clean left;
		}
		if (right != NIL) {
			clean right;
		}
	}
};


int (BinaryTree *b) addNode(int val) {
	if (val < b.val) {
		if (b.left == NIL) {
			b.left = make BinaryTree(val);
		} else {
			b.left.addNode(val);
		}
	} else if (val > b.val) {
		if (b.right == NIL) {
			b.right = make BinaryTree(val);
		} else {
			b.right.addNode(val);
		}
	} else {
		b.addNode(val+1);
	}
	return 0;
}

int (BinaryTree *b) find_max() {
	if (b.right == NIL) {
		return b.val;
	} else {
		return b.right.find_max();
	}
}

int (BinaryTree *b) find_min() {
	if (b.left == NIL) {
		return b.val;
	} else {
		return b.left.find_min();
	}
}

int main() {
	struct BinaryTree *b;
	int i;

	b = make BinaryTree(100);
	for (i = 0; i < 200; i++) {
		b.addNode(i);
	}
	b.find_max();
	clean b;
	return 0;
}
