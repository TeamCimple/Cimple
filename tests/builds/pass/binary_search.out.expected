has 200 0
hass 100 1
cleaning node 99
cleaning node 98
cleaning node 97
cleaning node 96
cleaning node 95
cleaning node 94
cleaning node 93
cleaning node 92
cleaning node 91
cleaning node 90
cleaning node 89
cleaning node 88
cleaning node 87
cleaning node 86
cleaning node 85
cleaning node 84
cleaning node 83
cleaning node 82
cleaning node 81
cleaning node 80
cleaning node 79
cleaning node 78
cleaning node 77
cleaning node 76
cleaning node 75
cleaning node 74
cleaning node 73
cleaning node 72
cleaning node 71
cleaning node 70
cleaning node 69
cleaning node 68
cleaning node 67
cleaning node 66
cleaning node 65
cleaning node 64
cleaning node 63
cleaning node 62
cleaning node 61
cleaning node 60
cleaning node 59
cleaning node 58
cleaning node 57
cleaning node 56
cleaning node 55
cleaning node 54
cleaning node 53
cleaning node 52
cleaning node 51
cleaning node 50
cleaning node 49
cleaning node 48
cleaning node 47
cleaning node 46
cleaning node 45
cleaning node 44
cleaning node 43
cleaning node 42
cleaning node 41
cleaning node 40
cleaning node 39
cleaning node 38
cleaning node 37
cleaning node 36
cleaning node 35
cleaning node 34
cleaning node 33
cleaning node 32
cleaning node 31
cleaning node 30
cleaning node 29
cleaning node 28
cleaning node 27
cleaning node 26
cleaning node 25
cleaning node 24
cleaning node 23
cleaning node 22
cleaning node 21
cleaning node 20
cleaning node 19
cleaning node 18
cleaning node 17
cleaning node 16
cleaning node 15
cleaning node 14
cleaning node 13
cleaning node 12
cleaning node 11
cleaning node 10
cleaning node 9
cleaning node 8
cleaning node 7
cleaning node 6
cleaning node 5
cleaning node 4
cleaning node 3
cleaning node 2
cleaning node 1
cleaning node 0
cleaning node 200
cleaning node 199
cleaning node 198
cleaning node 197
cleaning node 196
cleaning node 195
cleaning node 194
cleaning node 193
cleaning node 192
cleaning node 191
cleaning node 190
cleaning node 189
cleaning node 188
cleaning node 187
cleaning node 186
cleaning node 185
cleaning node 184
cleaning node 183
cleaning node 182
cleaning node 181
cleaning node 180
cleaning node 179
cleaning node 178
cleaning node 177
cleaning node 176
cleaning node 175
cleaning node 174
cleaning node 173
cleaning node 172
cleaning node 171
cleaning node 170
cleaning node 169
cleaning node 168
cleaning node 167
cleaning node 166
cleaning node 165
cleaning node 164
cleaning node 163
cleaning node 162
cleaning node 161
cleaning node 160
cleaning node 159
cleaning node 158
cleaning node 157
cleaning node 156
cleaning node 155
cleaning node 154
cleaning node 153
cleaning node 152
cleaning node 151
cleaning node 150
cleaning node 149
cleaning node 148
cleaning node 147
cleaning node 146
cleaning node 145
cleaning node 144
cleaning node 143
cleaning node 142
cleaning node 141
cleaning node 140
cleaning node 139
cleaning node 138
cleaning node 137
cleaning node 136
cleaning node 135
cleaning node 134
cleaning node 133
cleaning node 132
cleaning node 131
cleaning node 130
cleaning node 129
cleaning node 128
cleaning node 127
cleaning node 126
cleaning node 125
cleaning node 124
cleaning node 123
cleaning node 122
cleaning node 121
cleaning node 120
cleaning node 119
cleaning node 118
cleaning node 117
cleaning node 116
cleaning node 115
cleaning node 114
cleaning node 113
cleaning node 112
cleaning node 111
cleaning node 110
cleaning node 109
cleaning node 108
cleaning node 107
cleaning node 106
cleaning node 105
cleaning node 104
cleaning node 103
cleaning node 102
cleaning node 101
cleaning node 100
